library/psa_crypto.c

```diff
@@ -1342,6 +1365,229 @@ psa_status_t psa_export_key( mbedtls_svc_key_id_t key,
+                                            size_t *shared_secret_length )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    mbedtls_mpi P, G, X, GX, GY, K;
+    const size_t calculated_shared_secret_size = peer_key_length;
+
+    if( peer_key_length != our_key_length ||
+        calculated_shared_secret_size > shared_secret_size )
+    {
+        return(PSA_ERROR_INVALID_ARGUMENT);
+    }
+
+    mbedtls_mpi_init( &P ); mbedtls_mpi_init( &G );
+    mbedtls_mpi_init( &X ); mbedtls_mpi_init( &GX );
+    mbedtls_mpi_init( &GY ); mbedtls_mpi_init( &K );
+
+    status = psa_ffdh_set_prime_generator( peer_key_length, &P, &G );
+
+    if( status == PSA_SUCCESS )
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &X, our_key,
+                                                our_key_length ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &GY, peer_key,
+                                                peer_key_length ) );
+
+        /* Calculate our public key: GX = G^X mod P */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &GX, &G, &X, &P , NULL ) );
+
+        /* Calculate shared secret public key: K = G^(XY) mod P */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &K, &GY, &X, &P, NULL ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &K, shared_secret,
+                                                   calculated_shared_secret_size ) );
+
+        *shared_secret_length = calculated_shared_secret_size;
+    }
+cleanup:
+    mbedtls_mpi_free( &P ); mbedtls_mpi_free( &G );
+    mbedtls_mpi_free( &X ); mbedtls_mpi_free( &GX );
+    mbedtls_mpi_free( &GY ); mbedtls_mpi_free( &K );
+
+    if( status == PSA_SUCCESS && ret != 0 )
+        return( mbedtls_to_psa_error( ret ) );
+
+    return( PSA_SUCCESS );
+}
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_FFDH */
+
+static psa_status_t psa_export_ffdh_public_key_internal(
+    const uint8_t *key_buffer,
+    size_t key_buffer_size,
+    uint8_t *data,
+    size_t data_size,
+    size_t *data_length )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    mbedtls_mpi GX, G, X, P;
+
+    mbedtls_mpi_init( &GX ); mbedtls_mpi_init( &G );
+    mbedtls_mpi_init( &X ); mbedtls_mpi_init( &P );
+
+    status = psa_ffdh_set_prime_generator( key_buffer_size, &P, &G );
+
+    if( status == PSA_SUCCESS )
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &X, key_buffer,
+                                                key_buffer_size ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &GX, &G, &X, &P , NULL ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &GX, data, data_size ) );
+
+        *data_length = mbedtls_mpi_size( &GX );
+    }
+cleanup:
+    mbedtls_mpi_free( &P ); mbedtls_mpi_free( &G );
+    mbedtls_mpi_free( &X ); mbedtls_mpi_free( &GX );
+
+    if( status == PSA_SUCCESS && ret != 0 )
```

Comment:
function didn't change ret variable. So why is it checking ret variable here?
To me, it seems like this function will always return ERROR even if everything go correct.